File: api/function_app.js
// Azure Functions V2 (Node.js) — Science Converter backend.
// Same JSON contract as the Python version it replaces:
//   POST /api/convert  { "kind": "<key>", "value": <number> } -> { "result": <number> }
// Switched from Python to Node.js because Azure for Students does not currently
// expose the Linux + Python + Consumption combination in any region.

const { app } = require("@azure/functions");

const KM_PER_MILE = 1.609344;
const LB_PER_KG = 2.20462;

const CONVERTERS = {
  km_to_miles: (v) => v / KM_PER_MILE,
  miles_to_km: (v) => v * KM_PER_MILE,
  c_to_f:      (v) => (v * 9) / 5 + 32,
  f_to_c:      (v) => ((v - 32) * 5) / 9,
  kg_to_lb:    (v) => v * LB_PER_KG,
  lb_to_kg:    (v) => v / LB_PER_KG,
};

app.http("convert", {
  methods: ["POST", "OPTIONS"],
  authLevel: "anonymous",
  handler: async (request) => {
    if (request.method === "OPTIONS") {
      return { status: 204 };
    }

    let body;
    try {
      body = await request.json();
    } catch {
      return { status: 400, jsonBody: { error: "Request body must be JSON" } };
    }

    const { kind, value } = body || {};

    if (!CONVERTERS[kind]) {
      return { status: 400, jsonBody: { error: `Unknown conversion kind: ${kind}` } };
    }

    const numeric = Number(value);
    if (!Number.isFinite(numeric)) {
      return { status: 400, jsonBody: { error: "Field 'value' must be a number" } };
    }

    const result = CONVERTERS[kind](numeric);
    return { jsonBody: { kind, input: numeric, result } };
  },
});

app.http("health", {
  methods: ["GET"],
  authLevel: "anonymous",
  handler: async () => ({ jsonBody: { status: "ok" } }),
});

File: api/function_app.py
"""Azure Functions V2 backend for the Science Converter.

Single HTTP endpoint `/api/convert` that accepts JSON:
    { "kind": "<conversion key>", "value": <number> }
and returns:
    { "result": <number> }

The math is the same as converter.py in the repo root so the v1 in-browser
results and the v2 server-side results stay identical.
"""
import json
import azure.functions as func

KM_PER_MILE = 1.609344
LB_PER_KG = 2.20462

CONVERTERS = {
    "km_to_miles": lambda v: v / KM_PER_MILE,
    "miles_to_km": lambda v: v * KM_PER_MILE,
    "c_to_f":      lambda v: v * 9 / 5 + 32,
    "f_to_c":      lambda v: (v - 32) * 5 / 9,
    "kg_to_lb":    lambda v: v * LB_PER_KG,
    "lb_to_kg":    lambda v: v / LB_PER_KG,
}

app = func.FunctionApp(http_auth_level=func.AuthLevel.ANONYMOUS)


def _json(status: int, payload: dict) -> func.HttpResponse:
    return func.HttpResponse(
        body=json.dumps(payload),
        status_code=status,
        mimetype="application/json",
    )


@app.route(route="convert", methods=["POST", "OPTIONS"])
def convert(req: func.HttpRequest) -> func.HttpResponse:
    if req.method == "OPTIONS":
        return func.HttpResponse(status_code=204)

    try:
        body = req.get_json()
    except ValueError:
        return _json(400, {"error": "Request body must be JSON"})

    kind = body.get("kind")
    raw_value = body.get("value")

    if kind not in CONVERTERS:
        return _json(400, {"error": f"Unknown conversion kind: {kind}"})

    try:
        value = float(raw_value)
    except (TypeError, ValueError):
        return _json(400, {"error": "Field 'value' must be a number"})

    result = CONVERTERS[kind](value)
    return _json(200, {"kind": kind, "input": value, "result": result})


@app.route(route="health", methods=["GET"])
def health(_: func.HttpRequest) -> func.HttpResponse:
    return _json(200, {"status": "ok"})
